client/servers/manager.go

```diff
@@ -0,0 +1,529 @@
+			newServers = append(newServers, l.servers[i+1:]...)
+			l.servers = newServers
+
+			m.saveServerList(l)
+			return
+		}
+	}
+}
+
+// cycleServers returns a new list of servers that has dequeued the first
+// server and enqueued it at the end of the list.  cycleServers assumes the
+// caller is holding the listLock.  cycleServer does not test or ping
+// the next server inline.  cycleServer may be called when the environment
+// has just entered an unhealthy situation and blocking on a server test is
+// less desirable than just returning the next server in the firing line.  If
+// the next server fails, it will fail fast enough and cycleServer will be
+// called again.
+func (l *serverList) cycleServer() (servers []*Server) {
+	numServers := len(l.servers)
+	if numServers < 2 {
+		return servers // No action required
+	}
+
+	newServers := make([]*Server, 0, numServers)
+	newServers = append(newServers, l.servers[1:]...)
+	newServers = append(newServers, l.servers[0])
+
+	return newServers
+}
+
+// removeServerByKey performs an inline removal of the first matching server
+func (l *serverList) removeServerByKey(targetKey string) {
+	for i, s := range l.servers {
+		if targetKey == s.String() {
+			copy(l.servers[i:], l.servers[i+1:])
+			l.servers[len(l.servers)-1] = nil
+			l.servers = l.servers[:len(l.servers)-1]
+			return
+		}
+	}
+}
+
+// shuffleServers shuffles the server list in place
+func (l *serverList) shuffleServers() {
+	for i := len(l.servers) - 1; i > 0; i-- {
+		j := rand.Int31n(int32(i + 1))
+		l.servers[i], l.servers[j] = l.servers[j], l.servers[i]
+	}
+}
+
+// IsOffline checks to see if all the known servers have failed their ping
+// test during the last rebalance.
+func (m *Manager) IsOffline() bool {
+	offline := atomic.LoadInt32(&m.offline)
+	return offline == 1
+}
+
+// FindServer takes out an internal "read lock" and searches through the list
+// of servers to find a "healthy" server.  If the server is actually
+// unhealthy, we rely on Serf to detect this and remove the node from the
+// server list.  If the server at the front of the list has failed or fails
+// during an RPC call, it is rotated to the end of the list.  If there are no
+// servers available, return nil.
+func (m *Manager) FindServer() *Server {
+	l := m.getServerList()
+	numServers := len(l.servers)
+	if numServers == 0 {
+		m.logger.Printf("[WARN] manager: No servers available")
+		return nil
+	}
+
+	// Return whatever is at the front of the list because it is
+	// assumed to be the oldest in the server list (unless -
+	// hypothetically - the server list was rotated right after a
+	// server was added).
+	return l.servers[0]
+}
+
+// getServerList is a convenience method which hides the locking semantics
+// of atomic.Value from the caller.
+func (m *Manager) getServerList() serverList {
+	return m.listValue.Load().(serverList)
+}
+
+// saveServerList is a convenience method which hides the locking semantics
+// of atomic.Value from the caller.
```

Comment:
"hides the locking semantics of atomic.Value" is a weird phrase, but whatever. `getServerList` hides the type assertion nicely, and `saveServerList` makes a good companion api.